Move compute_tools from console repo (zenithdb/console#383)

ololobus opened the pull request:
Currently it's included with minimal changes and lives aside of the main
workspace. Later we may re-use and combine common parts with zenith
control_plane.

This change is mostly needed to unify cloud deployment pipeline:
1. build compute-tools image
2. build compute-node image based on the freshly built compute-tools
3. build zenith image

So we can roll new compute image and new storage required by it to
operate properly.

Resolve zenithdb/console#383

LizardWizzard reviewed (commented):
Great, thanks!

I've left some minor comments. Looks good to me in general.

I think that it is ok to add compute tools to workspace to be sure that it is picked up by all the tools, tests are discovered in CI etc.

Further improvements (can be done separately, probably it is better to create separate issue for that)

1. Logging utils from zenith_utils::logging can be reused
2. We have some postgresql config parsing routines in control_plane::postgresql_config so I think we can merge them with utils available in compute tools. 
3. Reuse basebackup/sync-safekeepers routines from control plane or replace control plane implementation with compute tools
3. Http building blocks from zenith_utils::http can be reused too

ololobus commented on compute_tools/src/bin/zenith_ctl.rs:
```diff
@@ -0,0 +1,264 @@
+
+    // Launch service threads first, so we were able to serve availability
+    // requests, while configuration is still in progress.
+    let mut _threads = vec![
```
Hm, join is blocking and those service threads shouldn't finish. But we do wait for postgres to exit and should exit as well after that immediately. So I don't know how join could be useful here

LizardWizzard replied:
> service threads shouldn't finish

Probably we dont care too much about clean shutdown here but maybe it is worth it so  there is no e.g broken connections/spurious errors during http requests (see http server shutdown in [zenith_utils](https://github.com/zenithdb/zenith/blob/main/zenith_utils/src/http/endpoint.rs#L184)).

ololobus commented on compute_tools/src/lib.rs:
```diff
@@ -0,0 +1,13 @@
+//!
+pub mod config;
+pub mod http_api;
+#[macro_use]
```
Nope, it is, rustc 1.57.0 (f1edd0429 2021-11-29):
```
error: cannot find macro `info_print` in this scope
   --> src/spec.rs:276:9
    |
276 |         info_print!("\n");
    |         ^^^^^^^^^^
    |
    = help: have you added the `#[macro_use]` on the module/import?
 ```

funbringer replied:
I think @LizardWizzard meant to say that you could just import this macro as a regular lang item.

ololobus commented on compute_tools/src/logger.rs:
```diff
@@ -0,0 +1,43 @@
+use chrono::Utc;
+use env_logger::{Builder, Env};
+
+macro_rules! info_println {
```
No :) Probably it's ugly, but that's for some pretty-printing. I use it to output some log message **with** date, thread name, etc. And then use these macro to output some 'graphics': roles / DBs list with actions on them. This is mostly for additional debug / info output, so I can live with plain info!, but these two are just more convenient, when I eyeball the logs. Once we switch to context-rich JSON logs, we can get rid of it, ofc

LizardWizzard replied:
IMO log with proper format settings and Debug/Display should do the trick. See how `tracing` is used e. g. in safekeeper where there are thread identifiers. So I do not see what is the difference here. log::info is the same thing just without boilerplate

ololobus commented on compute_tools/src/monitor.rs:
```diff
@@ -0,0 +1,107 @@
+
+                    // Connection is closed, reconnect and try again.
+                    client = Client::connect(&connstr, NoTls);
+                    continue;
```
Yeah, makes sense, thanks

ololobus replied:
On the second thought...
> Connection is established so looks like we can proceed with new connection
Do we? `Client::connect` can easily return error / closed connection. Later query will error out if it's closed, but I thought that it'd be just more clean to check whether we established a connection or not. And the easiest way to do it here without code duplication it just to continue

LizardWizzard replied:
> And the easiest way to do it here without code duplication it just to continue

If there is a function named something like `ensure_connected` which checks current connection and makes a new one if needed no code duplication is involved and all intentions are clean. Sometimes it is handy to have loop and it's iteration in different functions so you can use early returns and error propagation with `?`. This way we can eliminate extra delay between the error occurrence and a log message. (if these are far apart in logs this may complicate investigation when this is correlated from logs in different service)

ololobus commented on compute_tools/src/monitor.rs:
```diff
@@ -0,0 +1,107 @@
+                info!("cannot connect to postgres: {}, retrying", e);
+
+                // Establish a new connection and try again.
+                client = Client::connect(&connstr, NoTls);
```
Yep, but that's fine, IMO. Otherwise we should put sleep before every re-connect path and before every successful exit path

LizardWizzard replied:
Lets combine this with https://github.com/zenithdb/zenith/pull/1055/files#r775044231

ololobus commented on compute_tools/src/pg_helpers.rs:
```diff
@@ -0,0 +1,298 @@
+    }
+}
+
+pub trait PgOptionsSerialize {
```
GenericOptions is a type alias (or I don't know how it's properly called):
```rust
pub` type GenericOptions = Option<Vec<GenericOption>>;
```

and you cannot define impl on it. Or at least it wasn't possible, when I wrote it

LizardWizzard replied:
> type alias (or I don't know how it's properly called)

This is correct!

> and you cannot define impl on it

That's true for today's compiler too.

Another option is to use [newtype pattern](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). This creates a wrapper type and you can have an impl block on it.

If you find it less convenient feel free to stick with trait approach

ololobus commented on compute_tools/src/pg_helpers.rs:
```diff
@@ -0,0 +1,298 @@
+/// Execute `command` on the Postgres cluster using `pg_ctl` or `systemd` service.
+pub fn pg_service(pgdata: &Path, use_systemd: bool, command: &str) -> Result<()> {
+    if use_systemd {
+        Command::new("sudo")
```
No, IIRC, but that's the dead code, removed

LizardWizzard replied:
I briefly searched and there are some solutions to that, but I didnt check them myself

ololobus commented on compute_tools/src/pg_helpers.rs:
```diff
@@ -0,0 +1,298 @@
+            "-D",
+            &pgdata.to_path_buf().into_os_string().into_string().unwrap(),
+            "-l",
+            &log_path.into_os_string().into_string().unwrap(),
```
Wow, is it new? For some reason I haven't found it half a year back. This code is removed now

LizardWizzard replied:
Source [indicates](https://doc.rust-lang.org/src/std/path.rs.html#1957) that it is there since 1.0, maybe rustdoc didnt include that

ololobus commented on compute_tools/src/pg_helpers.rs:
```diff
@@ -0,0 +1,298 @@
+
+        if pid_path.exists() {
+            // XXX: dumb and the simplest way to get the last line in a text file
+            let stdout = Command::new("tail")
```
Put a TODO, cannot recall what kind of problems I had with it (I tried)

ololobus commented on compute_tools/src/spec.rs:
```diff
@@ -0,0 +1,280 @@
+            {
+                update_role = true;
+            } else if let Some(pg_pwd) = &r.encrypted_password {
+                update_role = pg_pwd[3..] != *role.encrypted_password.as_ref().unwrap();
```
Yes, sure, that's `md5:` trim

ololobus commented:
Fixed most of comments and put some TODOs for future. I removed the 100% dead code, but don't want to do a huge refactoring right now. I'll do another pass a bit later. Thanks a lot for the comments @LizardWizzard

funbringer commented on compute_tools/src/config.rs:
```diff
@@ -0,0 +1,47 @@
+use crate::pg_helpers::PgOptionsSerialize;
+use crate::zenith::ClusterSpec;
+
+/// Check that `line` is inside a text file and put it there if it is not.
```
Then why use regex?

ololobus replied:
Refactored it to do a one pass on a file and write at the end if needed

funbringer commented on compute_tools/src/bin/zenith_ctl.rs:
```diff
@@ -0,0 +1,263 @@
+        }
+    };
+
+    let compute_state = ComputeState {
+        connstr: connstr.to_string(),
+        pgdata: pgdata.to_string(),
+        pgbin: pgbin.to_string(),
+        spec,
+        ready: false,
+        last_active: Utc::now(),
+    };
+    let compute_state = Arc::new(RwLock::new(compute_state));
```
You don't have to wrap the whole struct in RwLock, because only `Utc` can't be updated trivially. I'd introduce the following changes:

```
ready: AtomicBool,
last_active: RwLock<Utc>, // Or even AtomicU64
```

ololobus replied:
My first intention was to use atomics here, but then I thought about:
- `ready` is likely to become a enum of at least (init, ready, running)
- Postgres child process handle is going to move into this struct, so we were able to update it's settings and send a SIGHUP signal, but we don't want to allow doing that concurrently
- We don't expect any real contention here

That way, it could be probably a combination of a couple of atomics + some sub-structure under mutex, but that just adds more complexity and I don't see any benefits now

funbringer replied:
> I don't see any benefits now

The benefit is that when you see `&ComputeState`, it's immediately obvious what can be changed. Rust isn't always about what's easier to write, but it gives us tools to make our intentions clear.

ololobus replied:
> The benefit is that when you see &ComputeState, it's immediately obvious what can be changed

Makes sense. Listed it for TODO here #1070

ololobus commented on compute_tools/src/bin/zenith_ctl.rs:
```diff
@@ -0,0 +1,263 @@
+            ),
+        }
+    }
+}
```
We used it with systemd previously and anyway it could be useful for local running / testing, so added PR_SET_PDEATHSIG, thanks.
